### PR TITLE
(NEW) ReplacedText has now been empowered with regex replacements

### LINKED ISSUE #534: ReplacedText empower with regex replacements
Let's make it possible to do this:

```java
String text = new ReplacedText(
  "Hello, my DEAR friend!",
  Pattern.compile("[A-Z]+"),
  matcher -> Integer.toString(matcher.group(1).length())
).asString();
```

It should produce:

```
"1ello, my 4 friend!"
```

Internally it should something like [this](https://stackoverflow.com/a/9606141/187141).

### PR DESCRIPTION
**Pull request by llorllale**
Closes #534

### CODE REVIEW

**Comment by 0crat** [comment]
@yegor256 please, pay attention to this pull request

**Comment by 0crat** [comment]
Job `gh:yegor256/cactoos#538` is now in scope, role is `REV`.

**Comment by 0crat** [comment]
This pull request [#538](https://github.com/yegor256/cactoos/issues/538) is assigned to [@fabriciofx](http://www.0crat.com/u/fabriciofx). The budget is 15 minutes, see [§4](http://datum.zerocracy.com/pages/policy.html#4). Please, read [§27](http://datum.zerocracy.com/pages/policy.html#27) and [§10](http://datum.zerocracy.com/pages/policy.html#10) and when you decide to accept the changes, inform [@yegor256](http://www.0crat.com/u/yegor256) right in this ticket.

File: src/main/java/org/cactoos/text/ReplacedText.java [hunk 1]
```diff
@@ -41,31 +45,76 @@
...
-        this.origin = text;
-        this.needle = find;
-        this.replacement = replace;
+        this(text, Pattern.compile(find), matcher -> replace);
     }
 
     @Override
     public String asString() throws IOException {
-        return this.origin.asString().replace(this.needle, this.replacement);
+        final StringBuffer buffer = new StringBuffer();
```
[the remark this answers is not in the document]
Comment (llorllale) [reply src/main/java/org/cactoos/text/ReplacedText.java]: @fabriciofx `Matcher.appendReplacement` began accepting `StringBuilder` in java 9. In java 8 we must use `StringBuffer`

File: src/main/java/org/cactoos/text/ReplacedText.java [hunk 2]
```diff
@@ -41,31 +45,76 @@
...
+     * @param regex The regular expression
+     * @param func Transforms the resulting matcher object into a replacement
+     *  string. Any exceptions will be wrapped in an {@link IOException}.
+     */
+    public ReplacedText(
+        final Text text,
+        final Pattern regex,
+        final Func<Matcher, String> func) {
+        this.origin = text;
+        this.regex = regex;
```
[the remark this answers is not in the document]
Comment (llorllale) [reply src/main/java/org/cactoos/text/ReplacedText.java]: @fabriciofx will do; however, have you guys considered using something like [HiddenField](http://checkstyle.sourceforge.net/config_coding.html#HiddenField) ?

[the remark this answers is not in the document]
Comment (fabriciofx) [reply src/main/java/org/cactoos/text/ReplacedText.java]: @llorllale I think so. But according with @yegor256 it's a bad code practice.

File: src/main/java/org/cactoos/text/ReplacedText.java [hunk 3]
```diff
@@ -47,13 +50,28 @@
...
-    private final Pattern regex;
+    private final Scalar<Pattern> regex;
 
     /**
      * The new replacement text.
      */
     private final Func<Matcher, String> replacement;
 
+    /**
+     * Shorthand ctor.
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/text/ReplacedText.java]: @llorllale Use just "Ctor."

File: src/main/java/org/cactoos/text/ReplacedText.java [hunk 4]
```diff
@@ -47,13 +50,28 @@
...
+     * <p>
+     * Will replace all instances of the substring matched by {@code find}
+     * with {@code replace}.
+     * @param text The text
+     * @param find The regular expression
+     * @param replace The replacement string
+     * @see #ReplacedText(Text, Pattern, Func)
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/text/ReplacedText.java]: @llorllale Remove this line. We usually use `@see` for reference to other classes.

File: src/main/java/org/cactoos/text/ReplacedText.java [hunk 5]
```diff
@@ -47,13 +50,28 @@
...
+     */
+    public ReplacedText(final Text text, final String find, final String
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/text/ReplacedText.java]: @llorllale Could you format this constructor, like the primary ctor?

File: src/main/java/org/cactoos/text/ReplacedText.java [hunk 6]
```diff
@@ -47,13 +50,28 @@
...
+        replace) {
+        this(text, () -> Pattern.compile(find), matcher -> replace);
+    }
+
     /**
      * Primary ctor.
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/text/ReplacedText.java]: @llorllale Same here. Just "Ctor."

File: src/main/java/org/cactoos/text/ReplacedText.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (fabriciofx) [reply src/main/java/org/cactoos/text/ReplacedText.java]: @llorllale Ok. I'll accept `StringBuffer`.

File: src/main/java/org/cactoos/text/ReplacedText.java [hunk 7]
```diff
@@ -41,31 +48,81 @@
...
     }
 
     @Override
     public String asString() throws IOException {
-        return this.origin.asString().replace(this.needle, this.replacement);
+        final StringBuffer buffer = new StringBuffer();
+        final Matcher matcher = new IoCheckedScalar<>(this.regex)
+            .value()
+            .matcher(this.origin.asString());
+        final IoCheckedFunc<Matcher, String> iofunc =
```
Comment (fabriciofx) [inline src/main/java/org/cactoos/text/ReplacedText.java]: @llorllale Please, rename the variable `iofunc` to `safe`.

**Comment by fabriciofx** [comment]
@yegor256 I accept this PR now. Could you merge it? Thanks!

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
Order was successfully finished: +15 points just awarded to @fabriciofx[/z](http://www.0crat.com/u/fabriciofx), total is [+255](http://www.0crat.com/u/fabriciofx)

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://datum.zerocracy.com/pages/policy.html#28): +15 points just awarded to @yegor256[/z](http://www.0crat.com/u/yegor256), total is [+7565](http://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
The job [#538](https://github.com/yegor256/cactoos/issues/538) is now out of scope

**Pull request merged**